### ISSUE TITLE
Default Layout Selection

### ISSUE THREAD

**Issue by hamishmack** [leksah/leksah#112]
Kuihm commented on reddit:

> I'd change the layout to make it more similar to VS, Eclipse and Intellij IDEA. Many people are used to have the project tree on their left and other windows at the bottom.

You can configure the layout described, but it is a bit fiddly.  I think we should have an easy way to reset the configuration to a selected default.

Defaults could be:

* Laptop Regular
* Laptop Goofy
* Desktop Regular
* Desktop Goofy
* Slab Regular
* Slab Goofy

What we default to now would be Laptop Goofy.  Regular would be a mirror image.  Desktop would be for larger screens where you can easily have three columns instead of two.  Slab would be for people who want to have stuff under the code window.

**Comment by asafeh** [comment]
One thing that would greatly improve the UI is the ability to easily collapse/drag the panes/tabs/notebooks/whatever with the mouse (akin to how visual studio and the like works). The current way of moving and rearranging the tabs is quite cumbersome and doesn't work very well (@jaccokrijnen opened the [issue](https://github.com/leksah/leksah/issues/193) I mentioned about crashs when dragging the tabs; that's the first thing I tried to do to save space on my small screen monitor at work).

Also, talking about usability, the "find menu" at the bottom could be reworked... Maybe separated into "Find", "Find in Files" and "Goto" options, according to what the user wants to do, something like Sublime Text does? It may sound trivial, but it's a bit confusing when you see it in the first time: "Replace" is an icon, "Replace all" is not, one might not instantly connect "Grep" with "Find in Files", etc.

Lastly, it might be just that I like distraction free settings, but everything in the layout seems to take more space than it should, the full screen option does not help at all.

Can I work on those suggestions? Is there something being thought in relation to UI?

**Comment by jaccokrijnen** [comment]
Good points, improving the usability is very important IMO.  I also had a "collapse pane group" feature on my idea list. Such that when two panes are part of a "split view", either one of them can collapse to only show the tabs of that group. Another related feature I had in mind is fixing the size of a panel in a split view. So that resizing the window does not resize both panes.
As for dragging, I think using a patched or previous GTK version would be the way to go, or did you have any other dragging related ideas?

Also, the "Initial Pane positions" editor (in the Preferences dialog) is quite clumsy to work with, I think it would be better for now to instead provide some sensible defaults like @hamishmack suggested and let the users pick from a dropdown menu or something like that.

As for the find bar I was thinking exactly the same thing :). I'd like the find bar to be only for the currently open file, and position it a bit better, maybe as a dropdown bar at the top of the editor. Then we could have a "Find in Files" dialog that also allows for some specific settings, e.g. selecting the folders to search, other grep options. By the way, I just noticed that grep isn't working on Windows. What do you think? And what do you mean by "Goto" options, by the way?

I'd say, go ahead and work on it :) (my current focus is on the Errors pane, proper background building and stack integration.). I saw that most of the pane/notebook related code is in the ltk package.

**Comment by hamishmack** [comment]
Added #196 for fixing grep on Windows.

**Comment by jaccokrijnen** [comment]
Related to the find bar: #178

**Comment by asafeh** [comment]
By goto I meant the option on the far right in the find bar. It feels out of place there.

I am still sorting out how things work to try and expiriment on those issues but I saw a couple things: 

* The unitial pane position in preferences is weird, and does not seem to work
* Setting to wrap lines and not show the margin while using the dark interface makes the window become larger than the screen (both windows 10 and fedora 22)

**Comment by jaccokrijnen** [comment]
Ah, I see. We could have an extra dropdown bar in the current editor for goto as well.

The initial panes settings is strange to work with: you have to select an entry in the list, then modify the text that is in the input fields below it, and then click "Add" to update it. 
I haven't found an easy way to insert custom widgets into the Preferences dialog, since it's based on the `ltk` library which works with an abstract description of an "editor" and then generates the widgets (https://github.com/leksah/leksah/blob/master/src/IDE/Preferences.hs#L538). So either it could be improved by extending ltk with a better "editor" for these kind of settings, or by having a list of defaults. (the same kind of editor is generated in the "Blacklist" and "Metadata" preferences)

I can't reproduce that second problem on Ubuntu, maybe open an issue?

**Comment by jaccokrijnen** [comment]
By the way, the "Default pane path" (which is used for editor tabs) is completely unusable. When the input in that field is invalid, the Apply button never becomes sensitive again (opened #201). Copy paste of valid input does the trick right now...

**Comment by asafeh** [comment]
I opened #204 for the window size issue.

The idea to have a list of default initial positions for the panes is horrible, IMO. If the user could easily show/hide/position the panes/menus/tabs, leksah would only need to have all available tabs/menus visible on the first run- then the user would arrange them to his liking and it'd be persisted in the IDE session. That's how the majority of the popular editors work.

**Comment by jaccokrijnen** [comment]
The pane layout is persisted with the leksah session file. The point is: where to open a tab when it's not open yet? That's what the defaults are for. 

> leksah would only need to have all available tabs/menus visible on the first run- 

I think that's not very nice, we shouldn't bother users with all possible panes at first startup. 

We could make an improvement by remembering the last path for each pane, only falling back on the default when it's never been opened before (but we'd have to think about the behaviour for opening documents) and also save the latest path with the .lkshs session file. Then we could perhaps omit the setting from Preferences at all, since it's not used very often.

Something else: The default setting is to store and load a session file per workspace. Do you think ( @hamishmack , @asafeh ) it would be better to use the global session file in ~/.leksah by default, or keep it like this? 
A benefit of the current setting is that users have their pane layout customized to the workspace, when opening a project you're back in the situation you left it (maybe with debugger open etc). This might also be a drawback (I like to have a workspace per-project and not have the panes shift when I change workspace). Maybe an in-between solution is to remember the open panes per session/workspace, but not the layout? (sizing, etc).
I prefer the global .lkshs by default.

**Comment by asafeh** [comment]
> The point is: where to open a tab when it's not open yet? That's what the defaults are for.

That still doesn't need a list of defaults. We could pick an arbitrary position for each pane, since rearragments will be persisted.

>I think that's not very nice, we shouldn't bother users with all possible panes at first startup.

I suggest showing them all because leksah doesn't have that many panes. Other possible solutions is to start with panes colapsed or show only the panes related to the state of the IDE, i.e.: if editing code show all panes about text files (browser,  etc), on debug show all debug related panes (debug, log, etc) and so on.

>Something else: The default setting is to store and load a session file per workspace. 

Do people really change workpaces? I also think a default workspace would be better; then we could make it transparent to the user: no nees to ask    where to save a workspace, just put it in ~/.leksah.

**Comment by jaccokrijnen** [comment]
>That still doesn't need a list of defaults. 

I agree, that's why I suggested:
>> We could make an improvement by remembering the last path for each pane, only falling back on the default when it's never been opened before (but we'd have to think about the behaviour for opening documents) and also save the latest path with the .lkshs session file. Then we could perhaps omit the setting from Preferences at all, since it's not used very often.

By persisting the last used pane path we could make things much easier.

> I suggest showing them all because leksah doesn't have that many panes. Other possible solutions is to start with panes colapsed or show only the panes related to the state of the IDE, i.e.: if editing code show all panes about text files (browser, etc), on debug show all debug related panes (debug, log, etc) and so on.

I agree about showing the relevant panes at per IDE state. But for startup, I think just showing the basic panes suffices.

> Do people really change workpaces?

At least I do :) Like I said:

> > I like to have a workspace per-project and not have the panes shift when I change workspace

I use workspaces similarly to Visual Studio's solutions: only include the relevant set of packages for my current project. This seems also more appropriate considering the planned stack support (multi-package projects).

**Comment by asafeh** [comment]
Ah, I see. I had a wrong idea of what the workspaces are used for. Visual Studio has a settings file for layout preferences and a sln file for each project. How is it done in leksah?

**Comment by jaccokrijnen** [comment]
leksah has .lkshw files for workspaces (set of packages), .lkshs files for "sessions" (pane state, stuff like opened files, full screen mode), and .lkhsp for preferences.

**Comment by asafeh** [comment]
Why not have a workspace per project?

**Comment by jaccokrijnen** [comment]
You mean one package per workspace instead of a set? When I'm working on Leksah for example, I have `leksah`, `ltk` and `leksah-server` added so I can easily jump between them.

**Comment by asafeh** [comment]
leksah, ltk and leksah-server could be a project, akin to VS solutions.

**Comment by jaccokrijnen** [comment]
In VS, a "solution" can have multiple "projects". In Leksah, a "workspace" can have multiple "packages". So leksah, ltk and leksah-server are packages in my "leksah" workspace (actually I have some other forked dependencies in there as well).

**Comment by asafeh** [comment]
Yes, that's what I talking about. So before you were asking whether the layout and position of the panes should be tied up to a workspace?

**Comment by jaccokrijnen** [comment]
Yes. Currently, a .lkshs session file is stored next to the workspace file unless you uncheck the box in Preferences -> Sessions. I don't really see an extra benefit in having more than one global session file for the pane configuration (I wouldn't want layout and stuff to change when I open another workspace, but maybe that's personal taste...).
Anyway, another option would be to uncheck that preference by default.

A different question: should "recent files" be stored in a session file? or not session-dependent, i.e. IDE global?

**Comment by asafeh** [comment]
Layout preferences should be global. I also don't see any advantage to the per workspace model.

For the recent files, I think either way is fine. Another option to think, though, is, on startup, open a start screen with a list of all workspaces instead of loading the last workspace.

**Comment by jaccokrijnen** [comment]
I've added #207, #208, #209 and #210

**Comment by asafeh** [comment]
@jaccokrijnen Is it possible for a panel to be on the left? It seems they group together with the code tabs if moved to the left.

**Comment by jaccokrijnen** [comment]
You mean a group of panes on the left?

You can build a setup of splits by going View -> Collapse (to collapse the split of the active panel) and View -> Split Horizontal or View -> Split Vertical to create new splits (I think the names "vertical" and "horizontal" are a bit confusing, I always pick the wrong ones...).

My current setup looks like this:

![leksah_setup](https://cloud.githubusercontent.com/assets/4022046/10045258/a9ea0292-6201-11e5-8a33-827fcf4fd265.png)
